[Tune] Nevergrad optimizer with extra parameters

yhna940 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Some Nevergrad search algorithms have required inputs, such as `budget` for the `NgOpt` search algorithm, but it is not possible with the NevergradSearch class to pass these parameters down to the search algorithm. I would propose adding something like optimizer_kwargs to the NevergradSearch that get passed to the optimizer when instantiating it.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number
#26305


<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

yhna940 commented:
> This pull request has been automatically marked as stale because it has not had recent activity. It will be closed in 14 days if no further activity occurs. Thank you for your contributions.
> 
> * If you'd like to keep this open, just leave any comment, and the stale label will be removed.

.

justinvyu commented on python/ray/tune/search/nevergrad/nevergrad_search.py:
```diff
@@ -120,6 +122,7 @@ def __init__(
         optimizer: Optional[
             Union[Optimizer, Type[Optimizer], ConfiguredOptimizer]
         ] = None,
+        optimizer_kwargs: Dict = dict(),
```
We can avoid the mutable default object here.

```suggestion
        optimizer_kwargs: Optional[Dict] = None,
```

justinvyu commented on python/ray/tune/search/nevergrad/nevergrad_search.py:
```diff
@@ -139,6 +142,7 @@ def __init__(
         self._space = None
         self._opt_factory = None
         self._nevergrad_opt = None
+        self._optimizer_kwargs = optimizer_kwargs
```
```suggestion
        self._optimizer_kwargs = optimizer_kwargs or {}
```

justinvyu commented on python/ray/tune/search/nevergrad/nevergrad_search.py:
```diff
@@ -166,6 +170,13 @@ def __init__(
                     "parameter."
                 )
+            if self._optimizer_kwargs:
+                raise ValueError(
```
Minor wording nit:

```python
raise ValueError(
    "If you pass in optimizer kwargs, either pass "
    "an `Optimizer` subclass or an instance of "
    "`ConfiguredOptimizer`."
)
```

yhna940 replied:
Changed the error message you suggested. thanks.

justinvyu commented on python/ray/tune/tests/test_searchers.py:
```diff
@@ -230,6 +230,26 @@ def testNevergrad(self):
+
+        out = tune.run(
+            _invalid_objective,
+            search_alg=NevergradSearch(
```
Can we limit the test to just the searcher initialization? Just making sure that initializing `NevergradSearch` with an optimizer that requires args without erroring should be good enough.

yhna940 replied:
I totally agree with you. Therefore, for this test case, only verify that nevergrad is instantiated. thanks.

justinvyu reviewed (changes requested):
Thanks for this contribution! I have a few minor suggestions, and a note about the failing test:

We should fix this test. Currently, it's passing in a `parameter_names` argument that's not being used at all and is being filled in as the `optimizer_kwargs`, causing the error. We should just remove this.

https://github.com/ray-project/ray/blob/a205fda91a1262dfb152918f70bf425d9ba5bca4/python/ray/tune/tests/test_tune_restore_warm_start.py#L253

Also, while we're at it, can we not accept `**kwargs` in the `NevergradSearch` constructor? It's currently not used at all and would create some confusion between the added `optimizer_kwargs` and the regular `kwargs`.

yhna940 commented:
> We should fix this test. Currently, it's passing in a `parameter_names` argument that's not being used at all and is being filled in as the `optimizer_kwargs`, causing the error. We should just remove this.
> 
> https://github.com/ray-project/ray/blob/a205fda91a1262dfb152918f70bf425d9ba5bca4/python/ray/tune/tests/test_tune_restore_warm_start.py#L253

Fixed by specifying `parameter_names`. I am so sorry.

yhna940 commented:
> Also, while we're at it, can we not accept `**kwargs` in the `NevergradSearch` constructor? It's currently not used at all and would create some confusion between the added `optimizer_kwargs` and the regular `kwargs`.

`**kwargs` is only used to call `__init__` of NevergradSearch's parent class.
The parameters used to build the optimizer used in `NevergradSearch` need to be defined separately.
As you said, `optimizer_kwargs` is too confusing, so we have replaced it with `optimizer_configs`.
Thank you so much for your review.

justinvyu commented on python/ray/tune/search/nevergrad/nevergrad_search.py:
```diff
@@ -52,6 +52,8 @@ class NevergradSearch(Searcher):
     Parameters:
         optimizer: Optimizer provided
             from Nevergrad. Alter
+        optimizer_configs: Dictionary used
+            to instantiate the Nevergrad optimizer.
```
```suggestion
        optimizer_configs: Kwargs passed in when instantiating the `optimizer`
```

justinvyu commented on python/ray/tune/search/nevergrad/nevergrad_search.py:
```diff
@@ -52,6 +52,8 @@ class NevergradSearch(Searcher):
     Parameters:
         optimizer: Optimizer provided
             from Nevergrad. Alter
```
```suggestion
        optimizer: Optimizer class provided from Nevergrad.
            See here for available optimizers:
            https://facebookresearch.github.io/nevergrad/optimizers_ref.html#optimizers
            This can also be an instance of a `ConfiguredOptimizer`. See the
            section on configured optimizers in the above link.
```

justinvyu commented on python/ray/tune/search/nevergrad/nevergrad_search.py:
```diff
@@ -120,6 +122,7 @@ def __init__(
         optimizer: Optional[
             Union[Optimizer, Type[Optimizer], ConfiguredOptimizer]
         ] = None,
+        optimizer_configs: Optional[Dict] = None,
```
Can we actually change this back to kwargs? We use this kind of naming in other parts of Tune.
```suggestion
        optimizer_kwargs: Optional[Dict] = None,
```

justinvyu commented on python/ray/tune/search/nevergrad/nevergrad_search.py:
```diff
@@ -139,6 +142,7 @@ def __init__(
         self._space = None
         self._opt_factory = None
         self._nevergrad_opt = None
+        self._optimizer_configs = optimizer_configs or {}
```
```suggestion
        self._optimizer_kwargs = optimizer_kwargs or {}
```

justinvyu commented on python/ray/tune/tests/test_searchers.py:
```diff
@@ -230,6 +230,12 @@ def testNevergrad(self):
         best_trial = out.best_trial
         self.assertLessEqual(best_trial.config["report"], 2.0)
 
+    def testNevergradWithConfigs(self):
```
```suggestion
    def testNevergradWithRequiredOptimizerKwargs(self):
```

justinvyu reviewed (changes requested):
Thanks!

In my comment early, I meant removing this `**kwargs` entirely and not passing it into the `super().__init__()`. The base `Searcher` class doesn't actually take in any args other than `metric` and `mode`, so it's not needed. Sorry for the confusion!

yhna940 commented:
> Thanks!
> 
> In my comment early, I meant removing this `**kwargs` entirely and not passing it into the `super().__init__()`. The base `Searcher` class doesn't actually take in any args other than `metric` and `mode`, so it's not needed. Sorry for the confusion!
> 
> Could you also merge with the latest master? (Make sure not to rebase! It will introduce a bunch of commits to this PR.)

Thank you for your review. There were 3 major changes.

1. Renamed back to `optimizer_kwargs`.

2. Merged with latest branch changes.

3. Removed kwargs in `super` `__init__` method.

justinvyu reviewed (approved):
Thank you for the contribution! I am retrying a few of the failed tests - they shouldn't be affected by these changes. Should be good to merge after they pass!

krfricke reviewed (approved):
Thanks!
